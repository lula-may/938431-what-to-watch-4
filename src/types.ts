export interface Review {
  author: string;
  authorId: number;
  date: Date;
  id: number;
  rating: number;
  text: string;
}

export interface Movie {
  actors: string[];
  bgColor: string;
  bgPoster: string;
  src: string;
  description: string;
  director: string;
  genre: string;
  id: number;
  isFavorite: boolean;
  poster: string;
  previewPoster: string;
  rating: {
    count: number,
    score: number,
  };
  releaseYear: number;
  runTime: number;
  title: string;
};
