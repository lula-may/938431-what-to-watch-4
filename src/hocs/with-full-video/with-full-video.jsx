import React, {PureComponent, createRef} from "react";
import PropTypes from "prop-types";
import {movieShape} from "../../components/shapes";

const withFullVideo = (Component) => {
  class WithFullVideo extends PureComponent {
    constructor(props) {
      super(props);

      this.state = {
        elapsedTime: 0,
        isLoading: true,
        isPlaying: false,
        progress: 0,
      };

      this._duration = 1;
      this._videoRef = createRef();

      this.handlePlayButtonClick = this.handlePlayButtonClick.bind(this);
      this.handleFullscreenButtonClick = this.handleFullscreenButtonClick.bind(this);
    }

    render() {
      const {elapsedTime, isPlaying, progress} = this.state;
      const {movie: {bigPoster}} = this.props;
      const progressValue = Math.round(progress * 100 / this._duration);
      return (
        <Component
          {...this.props}
          elapsedTime={elapsedTime}
          isPlaying={isPlaying}
          onFullScreenButtonClick={this.handleFullscreenButtonClick}
          onPlayButtonClick={this.handlePlayButtonClick}
          progressValue={progressValue}
        >
          <video
            className="player__video"
            ref={this._videoRef}
            poster={bigPoster}
          />
        </Component>
      );
    }

    componentDidMount() {
      const {movie: {src}} = this.props;
      const video = this._videoRef.current;
      video.src = src;

      video.onloadedmetadata = () => {
        this._duration = video.duration;
        this.setState({
          elapsedTime: this._duration,
          isLoading: false,
          isPlaying: true,
        });
      };

      video.ontimeupdate = () => {
        const progress = Math.floor(video.currentTime);
        const elapsedTime = (this._duration - progress);
        this.setState({
          progress,
          elapsedTime,
        });
      };

      video.onended = () => this.setState({
        isPlaying: false,
      });
    }

    componentDidUpdate() {
      const {isPlaying} = this.state;
      const video = this._videoRef.current;
      if (isPlaying) {
        video.play();
      } else {
        video.pause();
      }
    }

    componentWillUnmount() {
      const video = this._videoRef.current;
      video.onended = null;
      video.onloadedmetadata = null;
      video.ontimeupdate = null;
      video.poster = ``;
      video.src = ``;
    }

    handlePlayButtonClick() {
      const {isPlaying} = this.state;
      this.setState({
        isPlaying: !isPlaying
      });
    }

    handleFullscreenButtonClick() {
      const video = this._videoRef.current;

      if (!document.fullscreenElement) {
        video.requestFullscreen()
          .catch((err) => err);
      } else {
        document.exitFullscreen();
      }
    }
  }

  WithFullVideo.propTypes = {
    movie: PropTypes.shape(movieShape).isRequired,
  };

  return WithFullVideo;

};

export default withFullVideo;
