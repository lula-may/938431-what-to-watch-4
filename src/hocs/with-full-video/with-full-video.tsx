import * as React from "react";
import {connect} from "react-redux";
import {Subtract} from "utility-types";

import {Movie} from "../../types";
import {getMovieById} from "../../reducer/data/selectors";

const INITIAL_DURATION = 1;

interface Props {
  movie: Movie;
}

interface State {
  elapsedTime: number;
  isLoading: boolean;
  isPlaying: boolean;
  progress: number;
}

interface InjectedProps {
  elapsedTime: number;
  isPlaying: boolean;
  onFullScreenButtonClick: () => void;
  onPlayButtonClick: () => void;
  progressValue: number;
}

const withFullVideo = (Component) => {
  type P = React.ComponentProps<typeof Component>;
  type T = Props & Subtract<P, InjectedProps>;

  class WithFullVideo extends React.PureComponent<T, State> {
    private duration: number;
    private videoRef: React.RefObject<HTMLVideoElement>;

    constructor(props) {
      super(props);
      this.state = {
        elapsedTime: 0,
        isLoading: true,
        isPlaying: false,
        progress: 0,
      };
      this.duration = INITIAL_DURATION;
      this.videoRef = React.createRef();

      this.handlePlayButtonClick = this.handlePlayButtonClick.bind(this);
      this.handleFullscreenButtonClick = this.handleFullscreenButtonClick.bind(this);
    }

    handlePlayButtonClick() {
      const {isPlaying} = this.state;
      this.setState({
        isPlaying: !isPlaying
      });
    }

    handleFullscreenButtonClick() {
      const video = this.videoRef.current;
      if (!document.fullscreenElement) {
        video.requestFullscreen()
          .catch((err) => err);
      } else {
        document.exitFullscreen();
      }
    }

    componentDidMount() {
      const {movie: {src}} = this.props;

      const video = this.videoRef.current;
      video.src = src;
      video.onloadedmetadata = () => {
        this.duration = video.duration;
        this.setState({
          elapsedTime: this.duration,
          isLoading: false,
          isPlaying: true,
        });
      };

      video.ontimeupdate = () => {
        const progress = Math.floor(video.currentTime);
        const elapsedTime = (this.duration - progress);
        this.setState({
          progress,
          elapsedTime,
        });
      };

      video.onended = () => this.setState({
        isPlaying: false,
      });

      video.onfullscreenchange = () => {
        video.controls = (document.fullscreenElement && document.fullscreenElement.nodeName === `VIDEO`);
      };
    }

    componentDidUpdate() {
      const {isPlaying} = this.state;
      const video = this.videoRef.current;
      if (isPlaying) {
        video.play();
      } else {
        video.pause();
      }
    }

    componentWillUnmount() {
      const video = this.videoRef.current;
      video.onended = null;
      video.onloadedmetadata = null;
      video.ontimeupdate = null;
      video.poster = ``;
      video.src = ``;
    }

    render() {
      const {elapsedTime, isPlaying, progress} = this.state;
      const {movie: {bgPoster}} = this.props;
      const progressValue = Math.round(progress * 100 / this.duration);
      return (
        <Component
          {...this.props}
          elapsedTime={elapsedTime}
          isPlaying={isPlaying}
          onFullScreenButtonClick={this.handleFullscreenButtonClick}
          onPlayButtonClick={this.handlePlayButtonClick}
          progressValue={progressValue}
        >
          <video
            className="player__video"
            ref={this.videoRef}
            poster={bgPoster}
          />
        </Component>
      );
    }
  }

  return WithFullVideo;
};

const mapStateToProps = (state, props) => ({
  movie: getMovieById(state, props.match.params.id),
});

export {withFullVideo};

export default (Comp) => connect(mapStateToProps)(withFullVideo(Comp));
