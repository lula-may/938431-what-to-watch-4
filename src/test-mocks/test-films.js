const movieTitles = [
  `Gone With The Wind`,
  `The Godfather`,
  `Casablanca`,
  `A Streetcar Named Desire`,
  `Some Likes It Hot`,
];

const movies = movieTitles.map((title, i) => {
  return {
    title,
    id: i,
    src: `poster-${i}.jpg`
  };
});

const headerMovie = {
  title: `Jurassic Park`,
  genre: `Sci-Fi`,
  releaseYear: 1993,
  bg: `background.jpg`,
  poster: `header-poster.jpg`
};

export {movies, headerMovie};

