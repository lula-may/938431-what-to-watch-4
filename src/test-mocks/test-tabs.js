const tabs = [
  {
    id: `aa`,
    type: `First`
  },
  {
    id: `bb`,
    type: `Second`,
  },
  {
    id: `cc`,
    type: `Third`,
  }
];

export {tabs};
