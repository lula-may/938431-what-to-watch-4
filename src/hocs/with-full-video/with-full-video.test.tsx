import * as React from "react";
import * as renderer from "react-test-renderer";

import {withFullVideo} from "./with-full-video";
import {Movie} from "../../types";
import {testMovies} from "../../test-mocks/test-films";

const movie: Movie = testMovies[0];

interface Props {
  children: React.ReactNode;
}

const MockComponent: React.FC<Props> = (props: Props) => {
  const {children} = props;
  return (
    <div>
      {children}
    </div>
  );
};


const MockComponentWrapped = withFullVideo(MockComponent);

describe(`WithFullVideo HOC`, () => {
  it(`should render <video/> inside MockComponent`, () => {
    const tree = renderer.create(
        <MockComponentWrapped
          movie={movie}
        />, {
          createNodeMock: () => {
            return {};
          }
        }
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
