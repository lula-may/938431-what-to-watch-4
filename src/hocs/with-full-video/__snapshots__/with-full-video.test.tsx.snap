// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WithFullVideo HOC should render <video/> inside MockComponent 1`] = `
<div>
  <video
    className="player__video"
    poster="bip-poster-0.jpg"
  />
</div>
`;
